// apollo-server-cache-redis/dist/BaseRedisCache.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseRedisCache = void 0;
const dataloader_1 = __importDefault(require("dataloader"));
class BaseRedisCache {
    constructor(options) {
        this.defaultSetOptions = {
            ttl: 300,
        };
        const { client, noMgetClient } = options;
        if (client && noMgetClient) {
            throw Error('You may only provide one of `client` and `noMgetClient`');
        }
        else if (client) {
            this.client = client;
            this.loader = new dataloader_1.default((keys) => client.mget(...keys), {
                cache: false,
            });
        }
        else if (noMgetClient) {
            this.client = noMgetClient;
            this.loader = new dataloader_1.default((keys) => Promise.all(keys.map((key) => noMgetClient.get(key).catch(() => null))), {
                cache: false,
            });
        }
        else {
            throw Error('You must provide `client` or `noMgetClient`');
        }
    }
    set(key, value, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { ttl } = Object.assign({}, this.defaultSetOptions, options);
            if (typeof ttl === 'number') {
                yield this.client.set(key, value, 'EX', ttl);
            }
            else {
                yield this.client.set(key, value);
            }
        });
    }
    get(key) {
        return __awaiter(this, void 0, void 0, function* () {
            const reply = yield this.loader.load(key);
            if (reply !== null) {
                return reply;
            }
            return;
        });
    }
    delete(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.client.del(key)) > 0;
        });
    }
    flush() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.flushdb();
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.quit();
            return;
        });
    }
}
exports.BaseRedisCache = BaseRedisCache;
